File: algorithms/2_medium/M/00849_maximize_distance_to_closest_person/solution.cpp
typedef ptrdiff_t isize;

class Solution {
public:
    static int maxDistToClosest(const span<int> seats) {
        assert(any_of(seats.cbegin(), seats.cend(), [](const int a) { return a == 0; }));
        assert(any_of(seats.cbegin(), seats.cend(), [](const int a) { return a == 1; }));

        const isize len = std::ssize(seats);
        isize prev = -1;
        isize res = 0;
        for (isize i = 0; i < len; i += 1) {
            assert(seats[i] == 0 || seats[i] == 1);

            if (seats[i] == 1) {
                switch (prev) {
                    case -1:
                        res = i;
                        break;
                    default:
                        res = std::max(res, (i - prev) / 2);
                        break;
                }

                prev = i;
            }
        }
        res = std::max(res, len - 1 - prev);

        assert(res <= numeric_limits<int>::max());
        return static_cast<int>(res);
    }
};

File: algorithms/2_medium/M/00849_maximize_distance_to_closest_person/solution.rs
impl Solution {
    pub fn max_dist_to_closest(seats: Vec<i32>) -> i32 {
        assert!(seats.iter().any(|&seat| seat == 0));
        assert!(seats.iter().any(|&seat| seat == 1));

        let len = seats.len();
        let mut prev: Option<usize> = None;
        let mut res = 0_usize;
        for (i, seat) in seats.into_iter().enumerate() {
            assert!(seat == 0 || seat == 1);

            if seat == 1 {
                match prev {
                    None => res = i,
                    Some(other) => res = res.max((i - other) / 2),
                }
                prev = Some(i);
            }
        }
        res = res.max(len - 1 - prev.unwrap());

        res.try_into().unwrap()
    }
}
